### DISCUSSION TITLE
Centering CTk() and Tk() windows to  the main display (windows 10&11)

### DISCUSSION THREAD

**Discussion by HyperNylium** [TomSchimansky/CustomTkinter#1820] [Showcase]
When the CTk() window launches, it launches somewhere on your main display but doesn't really have a center function. I know this is kind of common knowledge by now but just wanted to put this here for anyone that is looking for it:
```python
from customtkinter import CTk

def CenterWindowToDisplay(Screen: CTk, width: int, height: int, scale_factor: float = 1.0):
    """Centers the window to the main display/monitor"""
    screen_width = Screen.winfo_screenwidth()
    screen_height = Screen.winfo_screenheight()
    x = int(((screen_width/2) - (width/2)) * scale_factor)
    y = int(((screen_height/2) - (height/1.5)) * scale_factor)
    return f"{width}x{height}+{x}+{y}"

window = CTk()

# will launch a 900x400 window in the center of the main screen.
window.geometry(CenterWindowToDisplay(window, 900, 400, window._get_window_scaling()))

window.mainloop()
```

I originally made the function in tkinter but when i started to use customtkinter, the window wasn't really launching in the middle of the display. It was always offset a little. This new version takes in account for the scale factor of your display as well (100%, 125%, 150%, etc).

This is what the tkinter version looks like:
```python
from tkinter import Tk

def CenterWindowToDisplay(Screen: Tk, width: int, height: int):
    """Centers the window to the main display/monitor"""
    screen_width = Screen.winfo_screenwidth()
    screen_height = Screen.winfo_screenheight()
    x = int((screen_width/2) - (width/2))
    y = int((screen_height/2) - (height/1.5))
    return f"{width}x{height}+{x}+{y}"

window = Tk()

# will launch a 900x400 window in the center of the main screen.
window.geometry(CenterWindowToDisplay(window, 900, 400)) 

window.mainloop()
```

You may use this code however you like :)

**Comment by jessielw** [comment]
Works perfect, thanks!

**Reply by HyperNylium** [reply]
No problem. Glad it helped!

**Reply by omegadev21** [reply]
i need if for a frame in custom tkinder not for a window but the code was help full for centering the window

**Reply by HyperNylium** [reply]
> i need if for a frame in custom tkinder not for a window but the code was help full for centering the window

Not sure if this would help you since you're saying you want to center a `CTkFrame` widget and not a window. The `CTkFrame` widget lives either on a `CTk()` window or a `Toplevel()` window as a widget. This function centers the CTk() window on your main screen.

If you are looking for a function to center a `Toplevel()` window to a `CTk()` window, I suggest you look [here](https://github.com/TomSchimansky/CustomTkinter/discussions/2510)

**Comment by magyrus1** [comment]
Funciona Perfeitamente
Obrigado ajudo muito D+.

**Comment by ghost** [comment]
Do you know how to do this with multiple monitors? Without manually changing some of the values

**Reply by HyperNylium** [reply]
Could you elaborate on what exactly you are looking for?

**Reply by beardedlinuxgeek** [reply]
I wanted it to open on my left monitor so I did `x -= screen_width`

**Reply by HyperNylium** [reply]
This function doesn't exactly do what you are looking for, sorry. This function was created for the sole purpose of launching a window exactly centered on your main display in the width and height you desired with the benefit of adjusting said width, height and launch position to the scale factor of the OS.

If you have a computer running Windows, this script/function defines "the main display" as whatever display you have selected as "Make this my main display" in **"Settings > System > Display > (select the display you want) > (tickbox) Make this my main display"**.
![image](https://github.com/TomSchimansky/CustomTkinter/assets/96999894/0844db11-eca4-4116-b2d0-d08053e6aaa9)

When a display is selected as the main display, the X and Y axis for that display become (0, 0). So on your monitor, the top left corner would be (0, 0). The first digit being the X coord and the second being the Y coord like this (x, y). But now let's say we are talking about your side monitor. Well, if it's on the left, then the numbers start to go into the negative, and if on the right the numbers are more positive. And by left or right, I am talking about the virtual display setup in Windows.
![image](https://github.com/TomSchimansky/CustomTkinter/assets/96999894/f5709850-962f-4d64-85ca-f3848f12e02e)

Mind you the (-1540, -1000) and (1540, 1000) change if your monitors are 720p, 1080p, 1440p, etc. I think they also change if the scale factor changes (125%, 150%, etc) but not 100% sure on that one. Made this function a long time ago 😅.

Anyways, to answer your question shortly, no, this function does not natively support centering CTk() or Tk() windows on different displays. As far as I remember `customtkinter.CTk().winfo_screenwidth()` and `customtkinter.CTk().winfo_screenheight()` return the X and Y coordinates of the main display, hence why I explained the whole main display in Windows thing above. 

I am not sure if there are libraries out there that support the kind of thing that you are looking for with the centering a window on different monitors but this was more of a built-in and simple way of centering a window, which is what I was looking for back in the day and made this function.

If you do find a different library that does what you are looking for or you decide to make your own script, please do share! I think the community would love a script/function like that :)

Hopefully this was helpful in some way :)

**Reply by ghost** [reply]
holy damn thanks for the detailed reply bro

**Reply by HyperNylium** [reply]
No problem :)

**Comment by Luppyn** [comment]
Apparently, there seems to be some kind of problem. On my 1920x1080 monitor, using practically the same function, I still get abruptly incorrect values. What could it be?

```python
def centralize(window_w:int, window_h:int, Screen:ct.CTk) -> str:
    print(window_w, window_h)
    return "{}x{}+{}+{}".format(
        window_w,
        window_h,
        ((Screen.winfo_screenwidth()/2)+(window_w/2)) * Screen._get_window_scaling(),
        ((Screen.winfo_screenheight()/2)+(window_h/1.5)) * Screen._get_window_scaling()
    )
```

![image](https://github.com/user-attachments/assets/2a3efcb2-415d-46d4-b10a-574e64b6259f)

**Reply by HyperNylium** [reply]
2 problems with this function:

1. In this line:
`((Screen.winfo_screenwidth()/2) + (window_w/2)) * Screen._get_window_scaling()`
That `+` needs to be a `-`.

2. You are not converting the final output to a integer before passing it into the `customtkinter.CTk().geometry()` function.

Your fixed function:
```python
def centralize(window_w:int, window_h:int, Screen:ct.CTk) -> str:
    return "{}x{}+{}+{}".format(
        window_w,
        window_h,
        int(((Screen.winfo_screenwidth()/2) - (window_w/2)) * Screen._get_window_scaling()),
        int(((Screen.winfo_screenheight()/2) - (window_h/1.5)) * Screen._get_window_scaling())
    )
```
